replica/table.cc

```diff
@@ -2149,14 +2174,14 @@ future<> tablet_storage_group_manager::handle_tablet_split_completion(const loca
             });
         }
         unsigned first_new_id = id << growth_factor;
-        auto split_ready_groups = std::move(*sg).split_ready_compaction_groups();
+        auto split_ready_groups = sg->split_ready_compaction_groups();
         if (split_ready_groups.size() != split_size) {
             on_internal_error(tlogger, format("Found {} split ready compaction groups, but expected {} instead.", split_ready_groups.size(), split_size));
         }
         for (unsigned i = 0; i < split_size; i++) {
             auto group_id = first_new_id + i;
             split_ready_groups[i]->update_id_and_range(group_id, new_tmap.get_token_range(locator::tablet_id(group_id)));
-            new_storage_groups[group_id] = std::make_unique<storage_group>(std::move(split_ready_groups[i]));
+            new_storage_groups[group_id] = make_lw_shared<storage_group>(std::move(split_ready_groups[i]));
```

Comment:
Is it ok that a loop can touch compaction groups which have their id and range updated?

Reply:
for example

let's say there are 10 tablets in this shard

there will be 30 compaction groups (left, main, and right groups for each tablet). left contains 1st half of range, main contains data spanning both halves, and right the 2nd half.

left, main and right share the same tablet id. showing we cannot rely on it for uniqueness.

all id and range tell you is which tablet id a group belongs to, and the token range that group spans, respectively.

I don't think a loop over groups need tablet id stability, since the loop is there for table-wide ops like major or flush. loops that need tablet id stability won't yield (see for_each_storage_group usage in tablet_sstable_set and split completion also must set all groups without yielding and that's what it does and must do always for many reasons).

And the range that we set during split completion will still be correct (before and after). we're just making it more fine-grained based on the new tablet count.